design/object-bucket/object-bucket.md

```diff
@@ -0,0 +1,148 @@
+- User: A Kubernetes cluster user with limited permissions, typically confined CRUD operations on Kubernetes objects within a single namespace.  The user will create request ceph object buckets by instantiating CephObjectBucketClaims. 
+
+## Use Cases
+
+### Assumptions
+
+1. A Rook-Ceph operator deployed on a Kubernetes Cluster.
+1. A CephObjectStore object exists along with the accompanying Service in a protected namespace.
+1. An existing CephObjectStoreUser exists along with the accompanying Secret in the user's namespace.
+
+##### Use Case: Expose an Object Store Endpoint for Bucket Provisioning
+
+As an admin, I want to expose an existing object store to cluster users so they can begin bucket provisioning via the Kubernetes API._
+
+![admin-actions](./cobc-admin.png)
+
+1. The admin creates a StorageClass with `parameters[serviceName]=objectstore-service` and `parameters[serviceNamespace]=some-namespace`
+  - _Note:_ the `provisioner` field is ignored.  Defining the provisioner is unnecessary as Rook-Ceph is the only operator watching for CephObjectBucketClaims.
+
+**Use Case: Provision a Bucket** 
+
+_As a Kubernetes user, I want to leverage the Kubernetes API to create Ceph Object Buckets. I expect to get back the bucket connection information in a Pod-attachable object._
+ 
+![user-actions](./cobc-user.png)
+
+0. The user creates a CephObjectBucketClaim in the user's namespace.
+1. The Rook Operator detects a new ObjectBucketClaim instance.  
+    1. The operator gets the user's credentials via `cephObjectBucketClaim.spec.secretName`.  
+    1. The operator gets the object store URL via `storageClass.parameters.serviceNamespace` + `storageClass.parameters.serviceName`.
+1. The operator uses the object store URL and credentials to make an S3 "make bucket" call.
+1. The operator creates a ConfigMap in the namespace of the CephObjectBucketClaim with relevant connection data for that bucket.
+1. An app Pod may then mount the Secret and the ConfigMap to begin accessing the bucket. 
+
+#####Use Case: Delete an Object Bucket
+
+_As a Kubernetes user, I want to delete ObjectBucketClaim instances and cleanup generated API resources._
+
+1. The user deletes the CephObjectBucketClaim via `kubectl delete ...`.
+1. The CephObjectBucketClaim is marked for deletion and left in the foreground.
+1. The respective ConfigMap is deleted.
+1. The CephObjectBucketClaim is garbage collected.
+
+---
+
+## API Specifications
+
+### Custom Resource Definition
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1beta1 
+kind: CustomResourceDefinition
+metadata:
+  name: cephobjectbucketclaims.rook.io
+spec:
+  group: ceph.rook.io
+  version: v1
+  scope: namespaced
+  names:
+      kind: CephObjectBucketClaim
+      listKind: CephObjectBucketClaimList
+      singular: cephobjectbucketclaim
+      plural: cephobjectbucketclaims
+```
+
+### Custom Resource
+
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: CephObjectBucketClaim
+metadata:
```

Comment:
do we want an OwnerReference pointing to the ceph cluster, so that the bucket claim is deleted automatically when the ceph cluster is deleted?

Reply:
I would say yes.  Will adjust.

Reply:
Agreed, we want every resource generated for the cluster to be cleaned up when the cluster is deleted.

Reply:
In this case, the OBC should not be a child resource since it's created by the user.